tests/std/tests/P0896R4_ranges_alg_transform/test.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <compare>
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
+
+constexpr void smoke_test() {

```

Comment:
In later tests - since the giant `test::range` rework PR - I've been merging the smoke test and instantiation test together to get more and better test coverage. I suggest you split this into three parts (and dump your existing instantiation tests):
1. The `STATIC_ASSERT`s can simply move into global scope.
2. unary transform test:
   ```c++
   struct unary_test {
       static constexpr std::array<P, 3> input      = {{{1, 99}, {4, 98}, {5, 97}}};
       static constexpr std::array<int, 3> expected = {{0, 3, 4}};

       template <class In, class Write>
       static constexpr void call() {
           // ...

           {
               // create an In wrapper for input and a Write wrapper for output for each transform call:
               In wrapped_in{input};
               auto result = transform(wrapped_in, Write{output}, minus_one, get_first);
               // ...
           }
       }
   };
   ```
  with corresponding `test_in_write<unary_transform_test, const P, int>()` in `main`.
3. binary transform test (similar, but with `In1` and `In2` template parameters.

We'll have to trim back the set of parameters used to instantiate the latter: `with_input_ranges` has 55 kinds of ranges and `with_writable_iterators` 15 varieties of iterator, so `test_in_in_write` will instantiate its argument 55 * 55 * 15 = 45375 times. See `ranges::search` for an example.

Reply:
I went with this approach and need to have a look at the more or less strange compile errors

Reply:
Done